core/embed/lib/image.c

```diff
@@ -224,35 +243,102 @@ secbool check_image_contents(const image_header *const hdr, uint32_t firstskip,
     return secfalse;
   }
 
-  const void *data =
-      flash_area_get_address(area, firstskip, IMAGE_CHUNK_SIZE - firstskip);
-  if (!data) {
-    return secfalse;
-  }
-  int remaining = hdr->codelen;
-  if (sectrue !=
-      check_single_hash(hdr->hashes, data,
-                        MIN(remaining, IMAGE_CHUNK_SIZE - firstskip))) {
-    return secfalse;
-  }
-
-  remaining -= IMAGE_CHUNK_SIZE - firstskip;
+  // Check the firmware integrity, calculate and compare hashes
+  size_t offset = firstskip;
+  size_t end_offset = offset + hdr->codelen;
 
-  int chunk = 1;
+  while (offset < end_offset) {
+    size_t bytes_to_check = MIN(IMAGE_CHUNK_SIZE - (offset % IMAGE_CHUNK_SIZE),
+                                end_offset - offset);
 
-  while (remaining > 0) {
-    data = flash_area_get_address(area, chunk * IMAGE_CHUNK_SIZE,
-                                  IMAGE_CHUNK_SIZE);
+    const void *data = flash_area_get_address(area, offset, bytes_to_check);
     if (!data) {
       return secfalse;
     }
-    if (sectrue != check_single_hash(hdr->hashes + chunk * 32, data,
-                                     MIN(remaining, IMAGE_CHUNK_SIZE))) {
+
+    size_t hash_offset = (offset / IMAGE_CHUNK_SIZE) * 32;
+    if (sectrue !=
+        check_single_hash(hdr->hashes + hash_offset, data, bytes_to_check)) {
       return secfalse;
     }
-    chunk++;
-    remaining -= IMAGE_CHUNK_SIZE;
+
+    offset += bytes_to_check;
+  }
+
+  // Check the padding to the end of the area
+  end_offset = flash_area_get_size(area);
+
+  if (offset < end_offset) {
+    // Use the first byte in the checked area as the expected padding byte
+    // Firmware is padded with 0xFF, while the bootloader itself is padded with
```

Comment:
to be super clear on this: bootloader is padded with zeros if it's installed via combined bin or bootloader update from firmware. When flashing bootloader separately via makefile for example, it will be padded with 0xFF too. 

We do allow here padding with any byte, but it only make sense for the images to be padded with 0x00 or 0xFF. Should we add a constraint?

Reply:
I've updated comment and added a check. see 1729331c329c80047115cd01d5d5c84116443a10

Reply:
the ensure is wrong here, just return secfalse.

i.e. this function is called at the start of bootloader, if sore some reason the padding is not correct, the bootloader shutdowns and the user has no way of erasing the firmware and uploading something valid

Reply:
yes, that's true. Fixed 1f6ac8c06319d379bc13a2e8ae1834627811ac7a